extension/parquet/reader/variant/variant_binary_decoder.cpp

```diff
@@ -0,0 +1,393 @@
+	ptr += length_in_bytes;
+	return result;
+}
+
+} // namespace
+
+VariantMetadataHeader VariantMetadataHeader::FromHeaderByte(uint8_t byte) {
+	VariantMetadataHeader header;
+	header.version = byte & VERSION_MASK;
+	header.sorted_strings = (byte >> SORTED_STRINGS_SHIFT) & SORTED_STRINGS_MASK;
+	header.offset_size = ((byte >> OFFSET_SIZE_MINUS_ONE_SHIFT) & OFFSET_SIZE_MINUS_ONE_MASK) + 1;
+
+	if (header.version != 1) {
+		throw NotImplementedException("Only version 1 of the Variant encoding scheme is supported, found version: %d",
+		                              header.version);
+	}
+
+	return header;
+}
+
+VariantMetadata::VariantMetadata(const string_t &metadata) : metadata(metadata) {
+	auto metadata_data = metadata.GetData();
+
+	header = VariantMetadataHeader::FromHeaderByte(metadata_data[0]);
+
+	const_data_ptr_t ptr = reinterpret_cast<const_data_ptr_t>(metadata_data + sizeof(uint8_t));
+	idx_t dictionary_size = ReadVariableLengthLittleEndian(header.offset_size, ptr);
+
+	offsets = ptr;
+	bytes = offsets + ((dictionary_size + 1) * header.offset_size);
+	idx_t last_offset = ReadVariableLengthLittleEndian(header.offset_size, ptr);
+	for (idx_t i = 0; i < dictionary_size; i++) {
+		auto next_offset = ReadVariableLengthLittleEndian(header.offset_size, ptr);
+		strings.emplace_back(reinterpret_cast<const char *>(bytes + last_offset), next_offset - last_offset);
+		last_offset = next_offset;
+	}
+}
+
+VariantValueMetadata VariantValueMetadata::FromHeaderByte(uint8_t byte) {
+	VariantValueMetadata result;
+	result.basic_type = VariantBasicTypeFromByte(byte & BASIC_TYPE_MASK);
+	uint8_t value_header = byte >> VALUE_HEADER_SHIFT;
+	switch (result.basic_type) {
+	case VariantBasicType::PRIMITIVE: {
+		result.primitive_type = VariantPrimitiveTypeFromByte(value_header);
+		break;
+	}
+	case VariantBasicType::SHORT_STRING: {
+		result.string_size = value_header;
+		break;
+	}
+	case VariantBasicType::OBJECT: {
+		result.field_offset_size = (value_header & FIELD_OFFSET_SIZE_MINUS_ONE_MASK) + 1;
+		result.field_id_size = ((value_header >> FIELD_ID_SIZE_MINUS_ONE_SHIFT) & FIELD_ID_SIZE_MINUS_ONE_MASK) + 1;
+		result.is_large = (value_header >> OBJECT_IS_LARGE_SHIFT) & OBJECT_IS_LARGE_MASK;
+		break;
+	}
+	case VariantBasicType::ARRAY: {
+		result.field_offset_size = (value_header & FIELD_OFFSET_SIZE_MINUS_ONE_MASK) + 1;
+		result.is_large = (value_header >> ARRAY_IS_LARGE_SHIFT) & ARRAY_IS_LARGE_MASK;
+		break;
+	}
+	default:
+		throw InternalException("VariantBasicType (%d) not handled", static_cast<uint8_t>(result.basic_type));
+	}
+	return result;
+}
+
+VariantBinaryDecoder::VariantBinaryDecoder(ClientContext &context) : context(context) {
+}
+
+template <class T>
+static T DecodeDecimal(const_data_ptr_t data, uint8_t &scale, uint8_t &width) {
+	scale = Load<uint8_t>(data);
+	data++;
+
+	auto result = Load<T>(data);
+	//! FIXME: The spec says:
+	//! The implied precision of a decimal value is `floor(log_10(val)) + 1`
```

Comment:
We're not setting the width according to the spec, but it doesn't seem to have an effect on the result.
I changed the implementation to respect this, but it made no difference